### PR TITLE
Allow name of column to be customized to support self-referencing many2many fields

### PR DESCRIPTION
**Pull request by nathan-osman**
This pull request attempts to solve issue #1194 without breaking anything else. This is accomplished by allowing the `AssociationForeignKey` tag to specify the name for the column in the database:

    type User struct {
        //...
        Friends []*User `gorm:"many2many:user_friends;AssociationForeignKey:ID=friend_id"`
    }

Notice the addition of the `=friend_id`. This addition is optional, which allows existing code to continue to work exactly the way it did before. Models that require a self-referencing many-to-many field can use the expanded syntax.

In `join_table_handler.go`, it became clear that `getSearchMap()` wasn't going to work in `Add()` since the `s.Source.ModelType == modelType` expression would evaluate to `true` for both the source and the destination models. Therefore, I adapted `Add()` to loop through the source and destination fields individually, ensuring that the correct scope was used.

The new changes also include a test.

### CODE REVIEW

File: model_struct.go [hunk 1]
```diff
@@ -289,11 +289,24 @@ func (scope *Scope) GetModelStruct() *ModelStruct {
...
+
+											// If a new name was provided for the field, use it
+											name = field.DBName
+											if len(parts) > 1 {
+												name = parts[1]
+											}
+
 											// join table foreign keys for association
-											joinTableDBName := ToDBName(elemType.Name()) + "_" + field.DBName
+											joinTableDBName := ToDBName(elemType.Name()) + "_" + name
```
Comment (jchavannes) [inline model_struct.go]: Why not just `name`?

Comment (nathan-osman) [reply model_struct.go]: This particular line will run even if an override is not specified. If I were to only assign `name` to `joinTableDBName`, then existing code would behave differently.

Comment (jchavannes) [reply model_struct.go]: Yes, but that can be changed.  If I set `ID=friend_id` then the column name should be `friend_id` not `user_friend_id`.

Comment (jchavannes) [reply model_struct.go]: I'm using your branch and made this change: https://github.com/jchavannes/gorm/commit/26342e

Also, thanks for creating this branch.  Self-referencing objects definitely needed to be supported.

Comment (nathan-osman) [reply model_struct.go]: Oh, I see what you mean. Yes, I'd be fine with doing it that way and I can easily update this pull request if that is the preferred behavior.

**Comment by shenshouer** [comment]
Any update for this ?

**Comment by jinzhu** [comment]
Hi @nathan-osman 

Merged it and changed the API a little bit, please refer http://gorm.io/associations.html#self-referencing-many-to-many-relationship for how to. 

Thank you for your awesome works.

**Comment by jchavannes** [comment]
@jinzhu as mentioned in my comment above, I much prefer the interface used in this branch otherwise all field names will always be prepended with the table name, which isn't ideal and loses flexibility: https://github.com/jchavannes/gorm/commit/26342e

**Comment by jinzhu** [comment]
@jchavannes GORM read name from tag `association_jointable_foreignkey` and won't prepend with table name. refer https://github.com/jinzhu/gorm/blob/gh-pages/documents/associations.md#self-referencing-many-to-many-relationship

**Comment by jchavannes** [comment]
Unless you changed something somewhere else, if you have `ID=friend_id` the column name in the database will be named `user_friend_id`, it is automatically prefixed by the table name.  It's caused by this line:

```
joinTableDBName := ToDBName(elemType.Name()) + "_" + name
```
https://github.com/jinzhu/gorm/pull/1709/files#diff-a608ea2eb85962d34d7b6e0bf6010af9R309

I've made a change that doesn't prepend the table name (linked in my previous comment).

Perhaps you prefer the columns being prefixed by the table name, but I'd rather have the flexibility to control that.

**Comment by jinzhu** [comment]
Hi @jchavannes 

Yes, I have changed the logic and its usage.

**Pull request closed without merging**